feat: [Spanner] make ValueMapper customizable

taka-oyama opened the pull request:
This PR makes Spanner's ValueMapper customizable.

I made this PR because I'd like to modify ValueMapper so that `decodeValues` return raw values instead of column types like TIMESTAMP being converted into `DateTimeImmutable` automatically. 
Right now, there is no good way to achieve this. 

I want this feature because in [Laravel](https://laravel.com/) (the web framework I use and [wrote a spanner driver for](https://github.com/colopl/laravel-spanner)), data is expected to be passed to the model as raw values and should be converted to useful types only when it is actually used, which is more efficient since not all rows retrieved from the database are actually used during the request's lifecycle.

This feature does not introduce any breaking changes.

Please let me know what you all think.

Thank you.

taka-oyama commented:
Rebased  to fix conflict.

CLA signing is going through the corporate pipeline and should be done by next week.

bshaffer commented:
Hi @taka-oyama ! Thanks for your contribution. This looks great, but I have a few questions.

> I made this PR because I'd like to modify ValueMapper so that decodeValues return raw values instead of column types like TIMESTAMP being converted into DateTimeImmutable automatically.

Is there a reason why you can't use `DateTimeImmutable` to convert back to the raw TIMESTAMP? Are there other examples of this?

taka-oyama commented:
Hi @bshaffer I could do that, and I plan on doing so if this doesn't work out, but I wanted to make an effort to make it run as optimal as possible, since we have had incidents in the past where unnecessary DateTime conversion was using up 10% of the total CPU in some projects.

> Are there other examples of this?

By "this", did you mean an example where a custom ValueMapper might be needed?

I haven't encountered any cases for any of the projects I've encountered, I think the same could be said for DATE type which is also converted to `DateTimeImmutable` and for BYTE type which are [decoded](https://github.com/googleapis/google-cloud-php/blob/main/Spanner/src/ValueMapper.php#L297) and converted into streams.

By the way, I noticed there was a similar request in google-cloud-go (https://github.com/googleapis/google-cloud-go/issues/854) and that was accepted.

bshaffer commented on Spanner/src/Database.php:
```diff
@@ -202,14 +204,15 @@ public function __construct(
         array $info = [],
-        $databaseRole = null
+        $databaseRole = null,
+        $mapper = null
```
let's name this variable `$valueMapper` for consistency

taka-oyama replied:
I've renamed all the variables and options from `$mapper` -> `$valueMapper`.

bshaffer commented on Spanner/src/Instance.php:
```diff
@@ -510,7 +510,8 @@ public function database($name, array $options = [])
             isset($options['database']) ? $options['database'] : [],
-            isset($options['databaseRole']) ? $options['databaseRole'] : null
+            isset($options['databaseRole']) ? $options['databaseRole'] : null,
+            isset($options['mapper']) ? $options['mapper'] : null
```
this needs to be documented as a valid entry for `$options` in the PHPDoc for this method

bshaffer commented on Spanner/tests/Unit/OperationTest.php:
```diff
@@ -287,6 +289,37 @@ public function testReadWithSnapshot()
+        $res = $operation->read($this->session, 'Posts', new KeySet(['all' => true]), ['foo']);
+        $this->assertInstanceOf(Result::class, $res);
+        $rows = iterator_to_array($res->rows());
+        $this->assertSame('99', $rows[0]['ID']);
```
This is failing in the tests right now (only on ubuntu, for some reason) with `Failed asserting that 10 is identical to '99'.`

bshaffer reviewed (commented):
@taka-oyama thank you for your prompt response!
I am shocked that DateTime processing would take up 10% of the total CPU, but perhaps processing thousands of datetime fields could have this effect? 

Thank you for providing the link to the Go PR as well, that is very helpful.

I'm placing a code sample here as well, so others can see how this option would be used in practice: 
```php
$spanner = new SpannerClient();
$instance = $spanner->instance('instance-name');
$database = $instance->database('database-name', ['valueMapper' => $myCustomDataMapper]);
```

taka-oyama commented:
Hi, sorry for the long wait.

The CLA is now passing.

> I am shocked that DateTime processing would take up 10% of the total CPU, but perhaps processing thousands of datetime fields could have this effect?

Yes, this happened when the code was processing a few thousand rows with many datetime columns.
Most frameworks just read columns with '*' by default, and it makes managing data easier but at the cost of performance.
This change will allow us to put less stress on performance.

bshaffer reviewed (commented):
This is LGTM from me, but I'd like to get @dwsupplee and @saranshdhingra's thoughts before merging!

vishwarajanand commented:
@taka-oyama thanks for raising this PR and fixing all outstanding. Upon discussion we are convinced about the reasoning for the feature requirement. However, we have reservations with the current changes, some of them are below:

1. `ValueMapperInterface` isn't very easy to implement for a developer as-is. It was meant to be an internal class and not exposed externally. Making it public will mean we will have to support it for back-compat (Google's other language libraries also don't seem to expose it). We are inclined to put this feature request in our backlog for whenever we revamp the library.

2. Alternatively, we want to do something like BigQuery's `returnRawResults` flag. [code pointer](https://github.com/googleapis/google-cloud-php/blob/7e53fe9c6e71e1a2321265a13299c734cbf575b4/BigQuery/src/BigQueryClient.php#L343-L345). And it will take a complete different implementation approach, so we could close this PR for now and move this to a feature request issue.

taka-oyama commented:
> Alternatively, we want to do something like BigQuery's returnRawResults flag. [code pointer](https://github.com/googleapis/google-cloud-php/blob/7e53fe9c6e71e1a2321265a13299c734cbf575b4/BigQuery/src/BigQueryClient.php#L343-L345). And it will take a complete different implementation approach, so we want to close this PR for now and move this to a feature request issue.

That works for me.

yash30201 commented:
Created a feature request issue for this (#5838) and thus closing this issue.
